sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala

```diff
@@ -773,15 +773,25 @@ class GpuHashAggregateIterator(
           // later. Cast here to the type that the aggregate expects (e.g. Long in case of count)
           val dataTypes = groupingExpressions.map(_.dataType) ++ aggregates.map(_.dataType)
 
-          val resCols = new mutable.ArrayBuffer[ColumnVector](result.getNumberOfColumns)
-          for (i <- 0 until result.getNumberOfColumns) {
-            val rapidsType = GpuColumnVector.getNonNestedRapidsType(dataTypes(i))
-            // cast will be cheap if type matches, only does refCount++ in that case
-            closeOnExcept(result.getColumn(i).castTo(rapidsType)) { castedCol =>
-              resCols += GpuColumnVector.from(castedCol, dataTypes(i))
+          val resCols = mutable.ArrayBuffer.empty[ColumnVector]
+          closeOnExcept(resCols) { resCols =>
+            (0 until result.getNumberOfColumns).foldLeft(resCols) { case (ret, i) =>
+              val column = result.getColumn(i)
+              val rapidsType = GpuColumnVector.getRapidsType(dataTypes(i))
+              // extra type conversion check for nested types
+              if ((rapidsType.equals(DType.LIST) || rapidsType.equals(DType.STRUCT)) &&
```

Comment:
Why not call `typeConversionAllowed` for all columns? (i.e. not need to special case LIST and STRUCT)?

Reply:
For non-nested types, it may happen type casting here, such as casting INT to LONG. Therefore, the check of `typeConversionAllowed` may fail when the type conversion is necessary.  For nested types, no type conversion is necessary (available), which indicates the check is safe.  What's more, we didn't match any children types in `GpuColumnVector.getRapidsType`. So, we check whether they match or not via `typeConversionAllowed`.